planner/core/memtable_predicate_extractor.go

```diff
@@ -662,6 +676,133 @@ func (e *ClusterLogTableExtractor) explainInfo(p *PhysicalMemTable) string {
 	return s
 }
 
+// HotRegionsHistoryTableExtractor is used to extract some predicates of `tidb_hot_regions_history`
+type HotRegionsHistoryTableExtractor struct {
+	extractHelper
+
+	// SkipRequest means the where clause always false, we don't need to request any pd server.
+	SkipRequest bool
+
+	// StartTime represents the beginning time of update time.
+	// e.g: SELECT * FROM tidb_hot_regions_history WHERE update_time>'2019-10-10 10:10:10.999'
+	StartTime int64
+	// EndTime represents the ending time of update time.
+	// e.g: SELECT * FROM tidb_hot_regions_history WHERE update_time<'2019-10-11 10:10:10.999'
+	EndTime int64
+
+	// RegionIDs/StoreIDs/PeerIDs represents all region/store/peer ids we should filter in PD to reduce network IO.
+	// e.g:
+	// 1. SELECT * FROM tidb_hot_regions_history WHERE region_id=1
+	// 2. SELECT * FROM tidb_hot_regions_history WHERE table_id in (11, 22)
+	// Leave range operation to above selection executor.
+	RegionIDs []uint64
+	StoreIDs  []uint64
+	PeerIDs   []uint64
+	// IsLearners/IsLeaders represents whether we should request for learner/leader role in PD to reduce network IO.
+	// e.g:
+	// 1. SELECT * FROM tidb_hot_regions_history WHERE is_learner=1
+	// 2. SELECT * FROM tidb_hot_regions_history WHERE is_learner in (0,1) -> request all
+	IsLearners []uint64
```

Comment:
Why do we only consider learners and leaders?

Reply:
is_learner and is_leader  are two fields of tidb_hot_regions_history, which is the same as  `information_schema.tikv_region_peers`. 
If user query `is_leader=1`, pd will only return leaders' hot regions. 
If user query `is_leader=0`, pd will only return followers' hot regions.
if user do not specify this field, executor will query both leader and followers' hot regions.
The detail logical is in `hotRegionsHistoryRetriver`:
```go
// Request all if no conditions
if len(e.extractor.IsLearners) == 0 {
e.extractor.IsLearners = []uint64{0, 1}
}
```
PD use a map struct to check the match sisuation of these push down predicates.

Reply:
The logic of is_learner is same as is_leader.